test/mod/test_buildroot.py

```diff
@@ -135,5 +143,41 @@ def test_selinuxfs_ro(self):
                    "/sys/fs/selinux",
                    "ro"]
 
-            r = root.run(cmd, monitor, readonly_binds=ro_binds)
+            r = root.run(cmd, monitor, timeout, readonly_binds=ro_binds)
             self.assertEqual(r.returncode, 0)
+
+    def test_timeout(self):
+        runner = detect_host_runner()
+        libdir = os.path.abspath(os.curdir)
+        var = pathlib.Path(self.tmp.name, "var")
+        var.mkdir()
+
+        with BuildRoot("/", runner, libdir, var) as root:
+
+            READ_ONLY = select.POLLIN | select.POLLPRI | select.POLLHUP | select.POLLERR
+            poller = select.poll()
+
+            proc = subprocess.Popen(['python3', libdir + '/stages/org.osbuild.test.timeout'], stdout = subprocess.PIPE)
+            start = time.monotonic()
+            timeout = 4
+            poller.register(proc.stdout.fileno(), READ_ONLY)
+            r = root.read_with_timeout(proc, poller, start, timeout)
+            self.assertEqual(r, b'Test stage output.\n')
+
+            proc = subprocess.Popen(['python3', libdir + '/stages/org.osbuild.test.timeout'], stdout = subprocess.PIPE)
+            start = time.monotonic()
+            timeout = 0
+            poller.register(proc.stdout.fileno(), READ_ONLY)
+            try:
+                r = root.read_with_timeout(proc, poller, start, timeout)
+                self.assertEqual(r, b'Test stage output.\n')
```

Comment:
This assert is only triggered if the timeout (of 0) is not hit (which it should be, but who knows about scheduling).

Reply:
With seeing the test stage, I understand this. But I think this should be improved to be stricter with the assertions:
- I would do one read to get the `Test stage output` string (not in a `try`)
- Then do the second read in a `with self.assertRaises(TimeoutError):` block